### PR TITLE
Put asset_events behind a run condition

### PR DESCRIPTION
**Pull request by james7132**
# Objective
Scheduling low cost systems has significant overhead due to task pool contention and the extra machinery to schedule and run them. Following the example of #7728, `asset_events` is good example of this kind of system, where there is no work to be done when there are no queued asset events.

## Solution
Put a run condition on it that checks if there are any queued events.

## Performance
Tested against `many_foxes`, we can see a slight improvement in the total time spent in `UpdateAssets`. Also noted much less volatility due to not being at the whim of the OS thread scheduler.
image

### CODE REVIEW

**Review by BorisBoutillier — APPROVED** [review]
Do we need to do a thorough pass on all bevy systems to check their internal complexity and if there is  a pertinent run condition to add, or is this still a compromise and always need to be checked against a benchmark ?

**Comment by james7132** [comment]
> Do we need to do a thorough pass on all bevy systems to check their internal complexity and if there is a pertinent run condition to add, or is this still a compromise and always need to be checked against a benchmark ?

The main goal here is performance so I'd always measure the results to ensure we don't have any regressions. Run conditions are always done singlethreadedly and block scheduling new system tasks, so there is a tradeoff to be made here when we're talking a large number of systems.

**Comment by hymm** [comment]
Is the Tracy with the system level spans disabled? That could significantly change the results.

**Comment by james7132** [comment]
No. I didn't disable system or run condition spans for either run. I'll go ahead and disable both and see how it goes.

File: crates/bevy_asset/src/assets.rs [hunk 1]
```diff
@@ -536,6 +536,14 @@ impl<A: Asset> Assets<A> {
     pub fn asset_events(mut assets: ResMut<Self>, mut events: EventWriter<AssetEvent<A>>) {
         events.send_batch(assets.queued_events.drain(..));
     }
+
+    /// A run condition for [`asset_events`]. The system will not run if there are no events to
+    /// flush.
+    ///
+    /// [`asset_events`]: Self::asset_events
+    pub fn asset_events_condition(assets: Res<Self>) -> bool {
```
Comment (alice-i-cecile) [inline crates/bevy_asset/src/assets.rs:544]: I don't think this needs to be `pub`.
```suggestion
    pub(crate) fn asset_events_condition(assets: Res<Self>) -> bool {
```

**Review by alice-i-cecile — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
I'm curious about those perf results, but won't block on them :)

**Comment by james7132** [comment]
> Is the Tracy with the system level spans disabled? That could significantly change the results.

image

Disabling the system spans makes the effect more pronounced. I disabled it both on this PR and main.

**Review by hymm — APPROVED** [review]
Thanks for redoing the benches.

**Pull request merged**